core/src/main/scala/chisel3/Bits.scala

```diff
@@ -148,6 +148,19 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
 
   /** @group SourceInfoTransformMacro */
   final def do_extract(x: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool = {
+    this.widthOption.foreach { thisWidth =>
+      if (thisWidth == 0) {
+        Builder.error(s"Cannot extract from zero-width $this")
+      } else {
+        x.widthOption.foreach { xWidth =>
+          if (xWidth >= 31 || (1 << (xWidth - 1)) >= thisWidth) {
+            Builder.warning(s"Dynamic index '$x' is too large for extractee '$this'")
+          } else if ((1 << xWidth) < thisWidth) {
+            Builder.warning(s"Dynamic index '$x' is too small for extractee '$this'")
+          }
+        }
+      }
+    }
     val theBits = this >> x
     theBits(0)
```

Comment:
```suggestion
    val noExtract = theBits.widthOption.exists(_ <= 1)
    if (noExtract) theBits else theBits(0)
```
> Is it fine to change the message back to just Cannot extract from zero-width so that they dedup?

No need to make them dedup, we can just elide the bit extract if it isn't useful and keep the more precise error message!